feat: Create a Growthbook server side provider

msamper opened the pull request:
<!-- Please use this template for your pull request. -->
<!-- Please use the sections that you need and delete other sections -->

## This PR
<!-- add the description of the PR here -->

- adds a new OpenFeature GrowthBook server provider

### How to test
Integrate OpenFeature into a test node server and set it up with this new GrowthBook provider. Ensure that GrowthBook flag evaluation works as expected using the OpenFeature API

[How to set up OpenFeature with node](https://openfeature.dev/docs/reference/technologies/server/javascript/)

OpenFeature.setProvider(new GrowthbookProvider(gbOptions, initOptions))

beeme1mr commented:
Hi @msamper, could you please rebase this PR? It should resolve the E2E issue.

msamper commented:
> Hi @msamper, could you please rebase this PR? It should resolve the E2E issue.

@beeme1mr I rebased on the main branch of my fork and it seems like there's still CI issues with the e2e and the lint-test-build. Should I be rebasing on [open-feature:main](https://github.com/open-feature/js-sdk-contrib/tree/main) instead?

lukas-reining reviewed (commented):
Thanks @msamper this looks good!
I just have two comments regarding the `EvaluationContext` -> `Attributes`.
When these are resolved this looks great!

lukas-reining commented:
> > Hi @msamper, could you please rebase this PR? It should resolve the E2E issue.
> 
> @beeme1mr I rebased on the main branch of my fork and it seems like there's still CI issues with the e2e and the lint-test-build. Should I be rebasing on [open-feature:main](https://github.com/open-feature/js-sdk-contrib/tree/main) instead?

Yes, [open-feature:main](https://github.com/open-feature/js-sdk-contrib/tree/main) should be your branch to rebase on.

msamper commented on libs/providers/growthbook/src/lib/growthbook-provider.ts:
```diff
@@ -0,0 +1,121 @@
+    defaultValue: boolean,
+    context: EvaluationContext,
+  ): Promise<ResolutionDetails<boolean>> {
+    if (!isEmpty(context)) {
+      // Add flag evaluation context and merge with existing context
+      await this.client.setAttributes({ ...this.client.getAttributes(), ...context });
+    }
```
That's right. The context would remain the same across both requests unless it was overwritten. I wasn't sure if a merge was needed to support the Dynamic Context Paradigm that is expected for server providers, but realizing that maybe the OpenFeature server SDK handles the different evaluation context levels before calling these functions.

lukas-reining replied:
That's true, you do not need to merge here, the SDK does it.

But this still has an issue.
If you do an evaluation with a context, the attributes of the GrowthBook object are set.
If another evaluation with absolutely no context is done after that, the attributes from the last evaluation will be used as they are not overwritten again.

This also relates to the concurrency problem I wrote in the other comment.

toddbaert commented on libs/providers/growthbook/src/lib/growthbook-provider.ts:
```diff
@@ -0,0 +1,121 @@
+    if (!isEmpty(context)) {
+      // Add flag evaluation context and merge with existing context
+      await this.client.setAttributes({ ...this.client.getAttributes(), ...context });
+    }
```
I agree. That pattern looks more like our [status context paradigm](https://openfeature.dev/docs/reference/concepts/sdk-paradigms#static-context-paradigms-client-side-sdks) we use for client.

Is there a way to set attributes without mutating the client in any way? @msamper ?

msamper replied:
As of right now our SDK doesn't have a way of doing that. We could consider making some changes to our SDK to support passing in some attributes that take precedence over ones within the GrowthBook context. I'm looking into it to see what the level of effort is and if there's some better alternatives

msamper commented:
We created a second version of our Node JS SDK that closely mirrors the Dynamic Context paradigm. I updated this PR to use it and fixed the tests. This should resolve the previously outstanding comments

msamper commented on libs/providers/growthbook/README.md:
````diff
@@ -0,0 +1,45 @@
+## Example Setup
+
+```typescript
+import { GrowthBook, Context, InitOptions } from '@growthbook/growthbook';
+import { GrowthbookProvider } from '@openfeature/growthbook-provider';
+import { OpenFeature } from '@openfeature/server-sdk';
+
+/*
+ * Configure your GrowthBook instance with GrowthBook context
+ * @see https://docs.growthbook.io/lib/js#step-1-configure-your-app
+ */
+const gbContext: Context = {
````
import `GrowthBookClient` instead and use `ClientOptions` not `Context`

msamper replied:
rename variables to reflect new types

beeme1mr reviewed (approved):
Looks good to me!

msamper commented:
> Looks good to me!

Great! Fixed up some variable names and updated the README so it should be good to merge
